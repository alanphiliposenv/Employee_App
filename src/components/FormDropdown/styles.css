.form-dropdown {
  display: flex;
  flex-direction: column;
}

select:invalid {
    color: gray;
}

.form-dropdown-select {
    margin-top: 8px;
    height: 45px;
    background: #FAFAFA 0% 0% no-repeat padding-box;
    border: 1px solid #D9D9D9;
    letter-spacing: 0px;
}
