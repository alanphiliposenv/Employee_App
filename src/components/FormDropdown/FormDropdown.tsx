import { FC } from 'react';
import './styles.css';

export type DropdownOptionsType = {
  label: string;
  value: string | number;
};

export type FormDropdownPropType = {
  placeholder?: string;
  label?: string;
  options: DropdownOptionsType[];
  value?: string | number;
  onChange?: (v) => void;
};

const FormDropdown: FC<FormDropdownPropType> = ({
  label,
  value,
  onChange,
  options,
  placeholder
}) => {
  return (
    <div className='form-dropdown'>
      <label data-testid='form-dropdown-label'>{label}</label>
      <select
        className='form-dropdown-select'
        value={value}
        style={value === '' || value === -1 ? { color: 'gray' } : { color: 'black' }}
        onChange={(e) => onChange(e.target.value)}
        data-testid='form-dropdown-select'
      >
        <option
          value={typeof value === 'number' ? -1 : ''}
          disabled
          hidden
          data-testid='form-dropdown-placeholder'
        >
          {placeholder}
        </option>
        {options.map((option) => {
          return (
            <option key={option.value} value={option.value} data-testid='form-dropdown-option'>
              {option.label}
            </option>
          );
        })}
      </select>
    </div>
  );
};

export default FormDropdown;
