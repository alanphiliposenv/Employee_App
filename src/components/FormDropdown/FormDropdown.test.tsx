import { fireEvent, render, screen } from '@testing-library/react';
import FormDropdown, { DropdownOptionsType, FormDropdownPropType } from './FormDropdown';
import '@testing-library/jest-dom';

describe('if FormDroppdown works properly', () => {
  test('if label rendered correctly', () => {
    const FormDropdownProps: FormDropdownPropType = {
      label: 'label-text',
      options: [] as DropdownOptionsType[],
      value: ''
    };

    render(<FormDropdown {...FormDropdownProps} />);
    const element = screen.getByTestId('form-dropdown-label');

    expect(element).toHaveTextContent('label-text');
  });
  test('if placeholder rendered correctly', () => {
    const FormDropdownProps: FormDropdownPropType = {
      placeholder: 'placeholder-text',
      options: [] as DropdownOptionsType[],
      value: ''
    };

    render(<FormDropdown {...FormDropdownProps} />);
    const element = screen.getByTestId('form-dropdown-placeholder');

    expect(element).toHaveTextContent('placeholder-text');
  });
  test('if options rendered correctly', () => {
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
      { label: 'label3', value: 'value3' },
      { label: 'label4', value: 'value4' }
    ];
    const FormDropdownProps: FormDropdownPropType = {
      options,
      value: ''
    };

    render(<FormDropdown {...FormDropdownProps} />);
    const elements = screen.getAllByTestId('form-dropdown-option');

    expect(elements.length).toBe(4);
    for (const opt of options) {
      const element = elements.find((element) => element.getAttribute('value') === opt.value);

      expect(element).toHaveTextContent(opt.label);
    }
  });
  test('if onChange is triggered', () => {
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
      { label: 'label3', value: 'value3' },
      { label: 'label4', value: 'value4' }
    ];
    const onChange = jest.fn();
    const FormDropdownProps: FormDropdownPropType = {
      options,
      onChange,
      value: ''
    };

    render(<FormDropdown {...FormDropdownProps} />);
    const element = screen.getByTestId('form-dropdown-select');

    fireEvent.change(element, { target: { value: 'value2' } });
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith('value2');
  });
  test('snapshot redered', () => {
    const options = [
      { label: 'label1', value: 'value1' },
      { label: 'label2', value: 'value2' },
      { label: 'label3', value: 'value3' },
      { label: 'label4', value: 'value4' }
    ];
    const FormDropdownProps: FormDropdownPropType = {
      options,
      value: ''
    };

    const element = render(<FormDropdown {...FormDropdownProps} />);

    expect(element).toMatchSnapshot();
  });
});
