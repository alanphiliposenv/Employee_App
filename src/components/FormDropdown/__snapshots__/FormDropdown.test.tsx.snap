// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if FormDroppdown works properly snapshot redered 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-dropdown"
      >
        <label
          data-testid="form-dropdown-label"
        />
        <select
          class="form-dropdown-select"
          data-testid="form-dropdown-select"
          style="color: gray;"
        >
          <option
            data-testid="form-dropdown-placeholder"
            disabled=""
            hidden=""
            value=""
          />
          <option
            data-testid="form-dropdown-option"
            value="value1"
          >
            label1
          </option>
          <option
            data-testid="form-dropdown-option"
            value="value2"
          >
            label2
          </option>
          <option
            data-testid="form-dropdown-option"
            value="value3"
          >
            label3
          </option>
          <option
            data-testid="form-dropdown-option"
            value="value4"
          >
            label4
          </option>
        </select>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-dropdown"
    >
      <label
        data-testid="form-dropdown-label"
      />
      <select
        class="form-dropdown-select"
        data-testid="form-dropdown-select"
        style="color: gray;"
      >
        <option
          data-testid="form-dropdown-placeholder"
          disabled=""
          hidden=""
          value=""
        />
        <option
          data-testid="form-dropdown-option"
          value="value1"
        >
          label1
        </option>
        <option
          data-testid="form-dropdown-option"
          value="value2"
        >
          label2
        </option>
        <option
          data-testid="form-dropdown-option"
          value="value3"
        >
          label3
        </option>
        <option
          data-testid="form-dropdown-option"
          value="value4"
        >
          label4
        </option>
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
